[tailwind] fix for postcss-import throwing errors when doing @import

dotnetCarpenter opened the pull request:
closing #2094

dotnetCarpenter committed:
[tailwind] fix for postcss-import throwing errors when doing @import

closing #2094

dotnetCarpenter committed:
add 'postcss-import' to postcss config

adamwathan commented on examples/tailwindcss/postcss.config.js:
```diff
@@ -1,6 +1,7 @@
 module.exports = {
   plugins: [
     require('tailwindcss')('./tailwind.js'),
+    require('postcss-import'),
```
It's better to run this plugin first so Tailwind can operate on a single set of CSS after the imports have been inlined, like mentioned in the `postcss-import` documentation:

> This plugin should probably be used as the first plugin of your list. This way, other plugins will work on the AST as if there were only a single file to process, and will probably work as you can expect.

If you tried to `@import` any CSS file that was using Tailwind features like `@responsive`, this setup wouldn't work because Tailwind wouldn't see that code, it would only see an `@import` line. You want all of the imports inlined before running any other plugins.

dotnetCarpenter replied:
You're right. I didn't see any difference, e.i. I didn't get any errors and my pages looked good. But I didn't check the output CSS.

adamwathan commented:
This PR is making assumptions that you aren't already using Sass, Less, or Stylus and that you want to use `postcss-import` instead which I worry makes it unnecessarily opinionated. We wouldn't make this change to the official Tailwind docs that have the same example file for example.

dotnetCarpenter commented:
All CSS in a standard `nuxt` project is run through PostCSS. Since the example doesn't setup Sass, Less or Stylus, I assume that it will run through PostCSS. Pretty sure I'm right on this one :)

dotnetCarpenter committed:
make sure that tailwind has access to the AST

It's better to run this plugin first so Tailwind can operate on a single set of CSS after the imports have been inlined, like mentioned in the postcss-import documentation:
> This plugin should probably be used as the first plugin of your list. This way, other plugins will work on the AST as if there were only a single file to process, and will probably work as you can expect.
If you tried to @import any CSS file that was using Tailwind features like @responsive, this setup wouldn't work because Tailwind wouldn't see that code, it would only see an @import line. You want all of the imports inlined before running any other plugins.

adamwathan commented:
*All* Tailwind projects run through PostCSS; that doesn't mean you aren't also using Less, Sass, or Stylus. We use Less for the Tailwind documentation for example, which outputs CSS, which is then fed through Tailwind using PostCSS.

Not every PostCSS user necessarily wants to use `postcss-import` at all which is why we didn't bake that right into Tailwind as a dependency.

dotnetCarpenter commented:
@adamwathan wow ok. I thought the choice was all or nothing. Hmm at least the comments should inform PostCSS users, which will be everyone, that if they want to use `postcss-import` they have to split `tailwind.css` up into seperate `.css` files before it will work.

Not sure if I should close this PR and just add a better description in the `tailwind.css` comments.

adamwathan commented:
This example code that's currently in the docs works perfectly fine without postcss-import for example:

```
/**
* This injects Tailwind's base styles, which is a combination of
* Normalize.css and some additional base styles.
*
* You can see the styles here:
* https://github.com/tailwindcss/tailwindcss/blob/master/css/preflight.css
*/
@tailwind preflight;

.btn { /* snip */ }
.form-input { /* snip */ }

/**
 * This injects all of Tailwind's utility classes, generated based on your
 * config file.
 */
@tailwind utilities;

.bg-pattern-graph-paper { /* snip */ }
.skew-45 { /* snip */ }
```

That's the only code we promise will work without a preprocessor; if you want to process `@import` statement in a preprocessor-like fashion using PostCSS, you can choose to add `postcss-import` to your build chain.

`@import` statements are totally valid in CSS as is (as long as they come first like the spec suggests); they don't always have to be inlined.

You can read more about the rules here:

https://developer.mozilla.org/en-US/docs/Web/CSS/%40import

The `@import` statements are preserved in the output normally, and the browser will make separate HTTP requests to fetch any files that are imported.

adamwathan commented:
So to be clear the reason I think it's a bad default to include `postcss-import` is because you take away the choice of whether or not to inline those imports from the user. If you add that plugin, you can't do a normal CSS import that does a separate HTTP request anymore.

dotnetCarpenter commented:
Thanks for clarifying. I think the following is very confusing:

```css
* Or if using a preprocessor:
*
* @import "components/buttons";
* @import "components/forms";
*/
```

Since PostCSS is a transpiler, which makes it a preprocessor in my book, not writing how to use PostCSS with tailwinds is very confusing. Can we agree to a different wording?

adamwathan commented:
I've updated our docs to mention that you need to import `@tailwind` calls from separate files if using postcss-import:

https://github.com/tailwindcss/tailwindcss/commit/4cf2cfa649c9988ca169f3d516491a5f30bcc979

This is really a postcss-import conversation though; it's not about "how to use Tailwind with PostCSS", it's about "how to inline import statements with PostCSS", to which the answer is `postcss-import`, and that plugin happens to be strict about the CSS spec so you can't just put imports anywhere you want, they have to be at the top of the file.

Folks have been discussing this exact error with that plugin long before Tailwind even existed 😊 

https://github.com/postcss/postcss-import/issues/257

dotnetCarpenter commented:
Thank you!

Just one question, why have you to switched to `@@` prefixes? I'm unaware of the syntax.

adamwathan commented:
The templating language we use for the documentation uses `@` to denote some of it's features, so we need to escape it with an extra `@`. It doesn't appear in the output:

![image](https://user-images.githubusercontent.com/4323180/32744016-35d1a570-c87c-11e7-81dd-9a55e9e33941.png)

dotnetCarpenter commented:
@adamwathan Thank you for being so attending. I have imported your comments into PR #2096

dotnetCarpenter commented:
Not sure that `postcss.config.js` is optimal for a nuxt project. It's hard to read the default postcss configuration since [style-loader.js](https://github.com/nuxt/nuxt.js/blob/dev/lib/builder/webpack/style-loader.js) has a vastly different structure.
